sweep/sweeper.go

```diff
@@ -880,52 +977,139 @@ func (s *UtxoSweeper) sweep(inputs inputSet,
 	return nil
 }
 
-// rescheduleInputs updates the pending inputs with the given tx inputs. It
-// increments the `publishAttempts` and calculates the next broadcast height
-// for each input. When the publishAttempts exceeds MaxSweepAttemps(10), this
-// input will be removed.
-func (s *UtxoSweeper) rescheduleInputs(inputs []*wire.TxIn) {
+// markInputsPendingPublish saves the sweeping tx to db and updates the pending
+// inputs with the given tx inputs. It also increments the `publishAttempts`.
+func (s *UtxoSweeper) markInputsPendingPublish(tr *TxRecord,
+	inputs []*wire.TxIn) error {
+
+	// Add tx to db before publication, so that we will always know that a
+	// spend by this tx is ours. Otherwise if the publish doesn't return,
+	// but did publish, we'd lose track of this tx. Even republication on
+	// startup doesn't prevent this, because that call returns a double
+	// spend error then and would also not add the hash to the store.
+	err := s.cfg.Store.StoreTx(tr)
+	if err != nil {
+		return fmt.Errorf("store tx: %w", err)
+	}
+
 	// Reschedule sweep.
 	for _, input := range inputs {
 		pi, ok := s.pendingInputs[input.PreviousOutPoint]
 		if !ok {
-			// It can be that the input has been removed because it
-			// exceed the maximum number of attempts in a previous
-			// input set. It could also be that this input is an
-			// additional wallet input that was attached. In that
-			// case there also isn't a pending input to update.
+			// It could be that this input is an additional wallet
+			// input that was attached. In that case there also
+			// isn't a pending input to update.
+			log.Debugf("Skipped marking input as pending "+
+				"published: %v not found in pending inputs",
+				input.PreviousOutPoint)
+
 			continue
 		}
 
+		// If this input has already terminated, there's clearly
+		// something wrong as it would have been removed. In this case
+		// we log an error and skip marking this input as pending
+		// publish.
```

Comment:
Is this truly an invariant that is supposed to never happen? If so, maybe we ought to panic in that block.

Reply:
I think my mental model is still as, we should avoid panic as much as possible - don't think this error should bring down the whole node.